kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -751,6 +751,24 @@ sealed abstract class Resource[F[_], +A] extends Serializable {
         K.combineK(allocate(this), allocate(that))
       }
 
+  /**
+   * A Resource where the acquire step is done lazily and memoized. This means that acquire
+   * happens only if and when the `F[A]` value is executed, instead of happening immediately
+   * upon `use()`. If the `F[A]` value is executed multiple times, acquire happens once only and
+   * the acquired resource is shared to all callers. The resource is released as normal at the
+   * end of `use` (whether normal termination, error, or cancelled), if it was acquired.
+   */
+  def memoizedAcquire[B >: A](implicit F: Concurrent[F]): Resource[F, F[B]] = {
+    Resource.eval(F.ref(List.empty[Resource.ExitCase => F[Unit]])).flatMap { release =>
+      val fa2 = F.uncancelable { poll =>
+        poll(allocatedCase).flatMap { case (a, r) => release.update(r :: _).as(a) }
+      }
+      Resource.makeCaseFull[F, F[B]](poll => poll(F.memoize(fa2).map(_.widen))) { (_, exit) =>
```

Comment:
```suggestion
      Resource.makeCaseFull[F, F[B]](poll => poll(F.memoize(fa2)).map(_.widen)) { (_, exit) =>
```

Otherwise I think it can get cancelled right after `F.memoize(fa2)`, and I think that would be not good.

Reply:
Oh I see, good catch. Probably we should add a test case.

Reply:
I will look into adding a test case for this, let's see what I can come up with.

Reply:
I tried to write a test for this but failed to find a reliable way to ensure acquisition is canceled right after the call to `F.memoize(fa2)` and before the end of `.map(_,widen)`. Is there some tool in the testkit that would allow me to do so and that I am not aware of?